### ISSUE TITLE
Allow for the Intercepters to be treated as a plugin (Pretender, MSW)

### ISSUE THREAD

**Issue by cah-briangantzler** [miragejs/miragejs#1037]
Currently the interceptor is Pretender and it is pretty hardcoded. The intent is to isolate the Interceptor code into its own configuration, then move it to its own package to allow for other interceptors to be specified.

Currently the main entry point to the interceptor code is on the server (the get, put, post, etc methods) and the server is the `this` then the routes are configured. The goal would be to move these methods to the interceptor code and off the server.

Tasks

- [ ] Move Pretender to its own config. Allow for a way to provide an alternate interceptor. Provide full backward compatibility and make Pretender the default interceptor. https://github.com/miragejs/miragejs/pull/1036
- [ ] Move the pretender code to its own package. (currently in progress see [mirage-pretender](https://github.com/bgantzler/mirage-pretender))
- [ ] Create a package that allows MSW to be used as an interceptor with Mirage. (Currently in progress see [mirage-msw](https://github.com/bgantzler/mirage-msw))
- [ ] Remove the Pretender code from MirageJS and require the user to choose the interceptor (Pretender, MSW, or none). This will be a breaking change in MirageJS.
- [ ] Move the interceptor creation methods (get, put, post, etc) to the interceptor config and make the interceptor config the `this` when the `routes` function is run. The Mirage Server will be supplied as a property on the config. For most users this should not be a breaking change, but more advanced uses this could be.

**Comment by samselikoff** [comment]
Plan seems good and in fact this was the road we were headed down in https://github.com/miragejs/miragejs/pull/699/files I believe... simplifying the main Mirage server api to be `server.handle(request)` and then letting folks do whatever they want with it.

Honestly it might be better to remove a hard-coded interceptor altogether and just let Mirage be concerned with turning a request into a response (and making sure it can run in both the browser and node). Then people can use pretender, MSW, even express.

**Comment by IanVS** [comment]
@samselikoff what do you think about making @cah-briangantzler a collaborator?  He's doing more work on this package than anyone else at this point, and while I'm happy to keep reviewing and helping out as I can, I don't want to be a bottleneck to the progress that Brian is making.

**Comment by cah-briangantzler** [comment]
Im actually a collaborator on ember-cli-mirage where this repo originated. The work I'm doing there has been spilling over into this one.  But I like to have another set of eyes, would prefer to not be the only one 😄

**Comment by henrymoulton** [comment]
@cah-brian-gantzler is this still active? Is MirageJS still active? I'm using it on a project and running into some issues.

**Comment by IanVS** [comment]
It's minimally active, yes.  The creator stepped away and a few of us have been trying to keep the lights on.  

MSW can indeed be used as an interceptor right now, using https://github.com/miragejs/mirage-msw.  It's a bit rough, though, since msw starts up asynchronously, and everything thus far in mirage has been sync.  So, to really get it working correctly we'll need to break some APIs I think.

**Comment by cah-brian-gantzler** [comment]
Yes, I created the initial support for MSW, but IanVS has been taking it to the next step improving it. Thanks IanVS.

**Comment by henrymoulton** [comment]
Thanks for the replies both, I'll take a look at moving away from Pretender.

**Comment by nickschot** [comment]
Since there's been some activity on this, I've been working on a `mirage-playwright` interceptor in a similar vein to support Mirage use in Playwright tests. Should hopefully be open source soon.

edit: https://github.com/qonto/mirage-playwright

**Comment by IanVS** [comment]
Status update: I've changed over my own non-trivial app to use mirage-msw, and it seems to be working well.  

There are new alpha versions of miragejs published in order to support the async-startup needs of msw.  So, if you want to try out mirage-msw, I recommend using at least the following versions:

- miragejs: `0.2.0-alpha.2`
- mirage-msw: `0.2.2`
- msw: `2.0.0`

It's still definitely experimental, but I think it's ready for a bit of alpha testing if anyone else is interested.

**Comment by iamkyrylo** [comment]
The industry is in an interesting situation: Mirage is moving toward adopting the MSW interceptor, while MSW lacks built-in mock data generation - something its users are increasingly expecting, especially since `@mswjs/data` is dead and [awaiting a rework](https://github.com/mswjs/data/issues/285#issuecomment-2455323230).

I think Mirage could benefit from this shift and gain more popularity if it extracted its db engine into a standalone library (e.g., `mirage-db`), making it accessible to MSW users. Just some thoughts...
